### PR TITLE
Implementation for the intern method

### PR DESCRIPTION
**Pull request by m3m0r7**

### CODE REVIEW

File: src/Kernel/Mnemonics/_if_acmpeq.php [hunk 1]
```diff
@@ -13,10 +15,14 @@ public function execute(): void
...
 
+        /**
+         * @var $rightOperand _String|_Utf8
+         * @var $leftOperand _String|_Utf8
+         */
         $rightOperand = $this->popFromOperandStack();
         $leftOperand = $this->popFromOperandStack();
 
-        if ($leftOperand === $rightOperand) {
+        if ($leftOperand->hashCode() === $rightOperand->hashCode()) {
```
Comment (chitoku-k) [inline src/Kernel/Mnemonics/_if_acmpeq.php]: Are you serious? In Java world,

```java
System.out.println("hello".hashCode() == new String("hello").hashCode());
```

prints true. If that is the case, please ignore this comment.

Comment (m3m0r7) [reply src/Kernel/Mnemonics/_if_acmpeq.php]: No, it is a trial implementation.

Comment (chitoku-k) [reply src/Kernel/Mnemonics/_if_acmpeq.php]: OK, thanks to have a look at this.

**Comment by m3m0r7** [comment]
@chitoku-k 

What do you think about this implementation?
Have any ideas?

https://github.com/php-java/php-java/pull/121/files#diff-8fdf42cf6b0a1e56ba8078c3880d1afdR18

https://github.com/php-java/php-java/blob/718a68cf00a99d919e8d2c97990887b3b06c4f6e/src/Utilities/Extractor.php#L18-L30

**Comment by m3m0r7** [comment]
@chitoku-k 
I fixed some implementation.

And, you need to make some changes as follows:

In `_String`:
```php
    /**
     * Returns a canonical representation for the string object.
     *
     * @native ConstantPool
     * @provider InternProvider
     * @param mixed $a
     * @return mixed
     * @throws NotImplementedException
     * @see https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/package-summary.html#intern
     */
    public function intern(ConstantPool $cp, InternProvider $internProvider)
    {
        foreach ($cp->getEntries() as $entry) {
            if (!($entry instanceof \PHPJava\Kernel\Structures\_Utf8)) {
                continue;
            }
            if ($entry->getString() === (string) $this) {
                $internProvider->get(StringIntern::class)[spl_object_id($this)] = $entry->getStringObject();
                break;
            }
        }
//        throw new NotImplementedException(__METHOD__);
    }
```

**Comment by chitoku-k** [comment]
@memory-agape Thanks for sharing! The API looks really much nicer and I can't wait for this to be merged. Please wait for a while until I go over the detail implementation.

File: src/Core/JVM/Intern/StringIntern.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<?php
+namespace PHPJava\Core\JVM\Intern;
+
+use PHPJava\Exceptions\NotAllowedNumberOfTypesException;
```
Comment (chitoku-k) [inline src/Core/JVM/Intern/StringIntern.php]: Remove this unused `use` statement.

File: src/Core/JavaClass.php [hunk 3]
```diff
@@ -15,6 +16,8 @@
 use PHPJava\Kernel\Attributes\AttributeInterface;
 use PHPJava\Kernel\Attributes\InnerClassesAttribute;
 use PHPJava\Kernel\Maps\FieldAccessFlag;
+use PHPJava\Kernel\Provider\DependencyInjectionProvider;
+use PHPJava\Kernel\Provider\InternProvider;
```
Comment (chitoku-k) [inline src/Core/JavaClass.php]: Needed here?

File: src/Kernel/Core/DependencyInjector.php [hunk 4]
```diff
@@ -5,43 +5,54 @@
...
+     * @param string $phpDocument
+     * @return array
+     * @throws NotSupportedInjectionTypesException
+     */
+    private function getNativeAnnotateInjections(string $phpDocument): array
     {
         $documentBlock = \phpDocumentor\Reflection\DocBlockFactory::createInstance()
             ->create($phpDocument);
 
         // Native annotation will dependency inject.
```
Comment (chitoku-k) [inline src/Kernel/Core/DependencyInjector.php]: “Native annotation will inject a dependency”.

File: src/Kernel/Core/DependencyInjector.php [hunk 5]
```diff
@@ -5,43 +5,54 @@
...
-                    case 'LocalStorages':
-                        $injections[] = $this->getLocalStorages();
-                        break;
-                    case 'Options':
-                        $injections[] = $this->options;
-                        break;
-                    default:
-                        throw new NotSupportedInjectionTypesException('Not supported injection type: "' . $type . '"');
-                }
+                $injections[] = $this->javaClassInvoker->getProvider(trim($provider));
```
Comment (chitoku-k) [inline src/Kernel/Core/DependencyInjector.php:55]: Looks much better, nice!

File: src/Kernel/Mnemonics/_astore_1.php [hunk 6]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace PHPJava\Kernel\Mnemonics;
 
+use PHPJava\Core\JVM\Intern\StringIntern;
```
Comment (chitoku-k) [inline src/Kernel/Mnemonics/_astore_1.php]: Needed here?

File: src/Kernel/Mnemonics/_astore_3.php [hunk 7]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace PHPJava\Kernel\Mnemonics;
 
+use PHPJava\Core\JVM\Intern\StringIntern;
```
Comment (chitoku-k) [inline src/Kernel/Mnemonics/_astore_3.php]: Needed here?

File: src/Kernel/Mnemonics/_if_acmpeq.php [hunk 8]
```diff
@@ -1,8 +1,12 @@
 <?php
 namespace PHPJava\Kernel\Mnemonics;
 
+use PHPJava\Core\JavaClass;
 use PHPJava\Exceptions\NotImplementedException;
+use PHPJava\Kernel\Structures\_String;
+use PHPJava\Kernel\Structures\_Utf8;
 use PHPJava\Utilities\BinaryTool;
+use PHPJava\Utilities\Extractor;
```
Comment (chitoku-k) [inline src/Kernel/Mnemonics/_if_acmpeq.php]: Needed here?

File: src/Kernel/Mnemonics/_if_acmpne.php [hunk 9]
```diff
@@ -1,8 +1,12 @@
 <?php
 namespace PHPJava\Kernel\Mnemonics;
 
+use PHPJava\Core\JavaClass;
 use PHPJava\Exceptions\NotImplementedException;
+use PHPJava\Kernel\Structures\_Utf8;
+use PHPJava\Packages\java\lang\_String;
 use PHPJava\Utilities\BinaryTool;
```
Comment (chitoku-k) [inline src/Kernel/Mnemonics/_if_acmpne.php]: Needed here?

File: src/Kernel/Structures/_Utf8.php [hunk 10]
```diff
@@ -12,19 +12,38 @@ class _Utf8 implements StructureInterface
...
+
+    public function __toString(): string
+    {
+        return $this->getString();
+    }
+
+    public function getStringObject(): \PHPJava\Packages\java\lang\_String
+    {
+        return $this->stringObject;
+    }
```
Comment (chitoku-k) [inline src/Kernel/Structures/_Utf8.php:48]: Awesome!

File: src/Utilities/Extractor.php [hunk 11]
```diff
@@ -1,11 +1,25 @@
 <?php
 namespace PHPJava\Utilities;
 
+use PHPJava\Kernel\Structures\_Utf8;
```
Comment (chitoku-k) [inline src/Utilities/Extractor.php]: Needed here?

**Comment by m3m0r7** [comment]
@chitoku-k 
Hey! I fixed tests and your reviews. 
Can you review again?

**Review by chitoku-k — APPROVED** [review]

**Comment by m3m0r7** [comment]
🐟 ("本気真鯵")

**Pull request merged**